packages/@aws-cdk/aws-secretsmanager/lib/secret.ts

```diff
@@ -0,0 +1,259 @@
+
+/**
+ * A secret in AWS Secrets Manager.
+ */
+export interface ISecret extends cdk.IConstruct {
+  /**
+   * The customer-managed encryption key that is used to encrypt this secret, if any. When ``unknown``, the default KMS
+   * key for the account and region is being used.
+   */
+  readonly encryptionKey?: kms.IEncryptionKey;
+
+  /**
+   * The ARN of the secret in AWS Secrets Manager.
+   */
+  readonly secretArn: string;
+
+  /**
+   * Exports this secret.
+   *
+   * @return import props that can be passed back to ``Secret.import``.
+   */
+  export(): SecretImportProps;
+
+  /**
+   * Grants reading the secret value to some role.
+   *
+   * @param grantee       the principal being granted permission.
+   * @param versionStages the version stages the grant is limited to. If not specified, no restriction on the version
+   *                      stages is applied.
+   */
+  grantRead(grantee: iam.IPrincipal, versionStages?: string[]): void;
+}
+
+/**
+ * The properties required to create a new secret in AWS Secrets Manager.
+ */
+export interface SecretProps {
+  /**
+   * An optional, human-friendly description of the secret.
+   */
+  description?: string;
+
+  /**
+   * The customer-managed encryption key to use for encrypting the secret value.
+   *
+   * @default a default KMS key for the account and region is used.
+   */
+  encryptionKey?: kms.IEncryptionKey;
+
+  /**
+   * The value of the secret. Either this or ``generateSecretString`` must be specified, but not both.
+   *
+   * @default a secret is generated using ``generateSecretString`` configuration.
+   */
+  secretString?: string;
+
+  /**
+   * Configuration for how to generate a secret value. Either this or ``secretString`` must be specified, but not both.
+   *
+   * @default the secret value specified in ``secretString`` is used.
+   */
+  generateSecretString?: SecretStringGenerator;
+
+  /**
+   * A name for the secret. Note that deleting secrets from SecretsManager does not happen immediately, but after a 7 to
+   * 30 days blackout period. During that period, it is not possible to create another secret that shares the same name.
+   *
+   * @default a name is generated by CloudFormation.
+   */
+  name?: string;
+}
+
+/**
+ * Attributes required to import an existing secret into the Stack.
+ */
+export interface SecretImportProps {
+  /**
+   * The encryption key that is used to encrypt the secret, unless the default SecretsManager key is used.
+   */
+  encryptionKey?: kms.IEncryptionKey;
+
+  /**
+   * The ARN of the secret in SecretsManager.
+   */
+  secretArn: string;
+}
+
+/**
+ * The common behavior of Secrets. Users should not use this class directly, and instead use ``Secret``.
+ */
+export abstract class SecretBase extends cdk.Construct implements ISecret {
+  public abstract readonly encryptionKey?: kms.IEncryptionKey;
+  public abstract readonly secretArn: string;
+
+  public export(): SecretImportProps {
```

Comment:
`export()` should be abstract in `Base` and implemented in `ImportedSecret` (see other examples)

Reply:
What does it buy me?

Reply:
It doesn't make sense to support `export` at the "Base" class because it doesn't know how the resource is manifested. See [guidelines](https://github.com/awslabs/aws-cdk/blob/master/design/aws-guidelines.md#exports)

Reply:
Right I guess I thought of the `Base` class as an implementation detail, but someone could make their own implementation that behaves different...